Request #714: Apply indirect=True on particular argnames

Elizaveta239 opened the pull request:
The results of EuroPython 2015 sprint. See #714.

The-Compiler commented on _pytest/python.py:
```diff
@@ -889,13 +892,22 @@ def parametrize(self, argnames, argvalues, indirect=False, ids=None,
-        valtype = indirect and "params" or "funcargs"
+        else:
+            if not isinstance(indirect, (tuple, list)):
+                valtypes = {arg: "params" for arg in argnames}
```
It seems dict comprehensions are invalid syntax in Python 2.6, which is why the checks are failing. You can probably do something like `dict([(arg, "params") for arg in argnames])` instead (but I didn't test it :wink:).

The-Compiler replied:
Ah, nevermind - you were faster! :laughing:

nicoddemus commented on testing/python/metafunc.py:
```diff
@@ -220,6 +220,46 @@ def func(x, y): pass
+        testdir.makepyfile("""
+            def pytest_generate_tests(metafunc):
+                metafunc.parametrize('x, y', [('a', 'b')], indirect=['x'])
+            def pytest_funcarg__x(request):
```
Minor, but I think using `@pytest.fixture` here is better as it is the modern idiom. :smile:

nicoddemus commented:
Thanks for this @Elizaveta239! Could you please also add a note the the `CHANGELOG` and add yourself to `AUTHORS`?

Elizaveta239 commented:
Hi @nicoddemus! Could you help me please? I added a test for bad 'indirect' parameter in order to increase code coverage. But this test isn't marked as 'passed', but marked as 'error'. How should I change it?

nicoddemus commented:
Sure thing, can you post the failing test?

Also, I think this feature deserves some documentation on it... could please add that too? :smile:

Elizaveta239 commented:
@nicoddemus I said about 'test_parametrize_indirect_list_error' in metafunc.py. But now I figured out that the problem appeared because of my local changes. Sorry for disinformation! :)
I want to update documentation, but I don't know how to do it.

nicoddemus commented:
I guess it is ready now, right? :smile: 

@pfctdayelise, could you also take a look?

Elizaveta239 commented:
Not yet, I'm going to update documentation soon :)

pfctdayelise commented on _pytest/python.py:
```diff
@@ -889,13 +892,22 @@ def parametrize(self, argnames, argvalues, indirect=False, ids=None,
         if scope is None:
             scope = "function"
         scopenum = scopes.index(scope)
+        valtypes = dict.fromkeys(argnames, "funcargs")
```
This section needs to be tidied up a bit. I think it would be simpler if you first deal with setting the right values in the valtypes dict, based on the value of indirect.
Then, secondly, go through that dict and check if we need to raise the "uses no fixture" exception.
Also, rather than checking if indirect is a tuple or list, it might be nicer to explicitly check for True and False, else assume it is a sequence and iterate over it.

pfctdayelise commented on testing/python/metafunc.py:
```diff
@@ -220,6 +220,53 @@ def func(x, y): pass
+        def func(x, y): pass
+        metafunc = self.Metafunc(func)
+        metafunc.parametrize('x, y', [('a', 'b')], indirect=['x'])
+        assert metafunc._calls[0].funcargs == dict(y='b')
```
This is not your problem or mistake, but it seems a bit yuck that _calls is protected...

pfctdayelise commented on testing/python/metafunc.py:
```diff
@@ -220,6 +220,53 @@ def func(x, y): pass
         assert metafunc._calls[0].params == dict(x=1,y=2, unnamed=1)
         assert metafunc._calls[1].params == dict(x=1,y=3, unnamed=1)
 
+    def test_parametrize_indirect_list(self):
```
Please add a docstring or in the name of the function, a reference to the issue number

pfctdayelise commented on testing/python/metafunc.py:
```diff
@@ -220,6 +220,53 @@ def func(x, y): pass
+        assert metafunc._calls[0].funcargs == dict(x='a', y='b')
+        assert metafunc._calls[0].params == {}
+
+    def test_parametrize_indirect_list_functional(self, testdir):
```
Nice test! It deserves a docstring so Future Us can understand why you've written it this way :)

pfctdayelise commented on testing/python/metafunc.py:
```diff
@@ -220,6 +220,53 @@ def func(x, y): pass
+    def test_parametrize_indirect_list_error(self, testdir):
+        def func(x, y): pass
+        metafunc = self.Metafunc(func)
+        pytest.raises(ValueError, lambda:
```
Did you know you can use pytest.raises as a context manager? It's a bit nicer IMO but that's just a style preference I guess

pfctdayelise commented on _pytest/python.py:
```diff
@@ -889,13 +892,22 @@ def parametrize(self, argnames, argvalues, indirect=False, ids=None,
+            else:
+                for arg in indirect:
+                    if arg not in argnames:
+                        raise ValueError("indirect: fixture %r doesn't exist" %(
```
Following the ValueError example above, I think it is useful to put the function name in the error message

Elizaveta239 commented:
The fixes after code review and docs are ready :)

pfctdayelise commented:
OK @Elizaveta239 I made some changes to your branch, but now I am not sure how to get them back to you! :worried: anyway they are in 40fa7b2 (merged up to date, and fixed merge conflict in CHANGELOG) and 36b86af (added test cases).

In the test cases I added two which are failing but which should be passing. Do you want to see if you can get them passing? If not, I can do it, but you're very close to completing this :)

Elizaveta239 commented:
@pfctdayelise Thank you! But I will not be able to do it during next two weeks :( If it's possible to wait until then -- I will fix these tests.

pfctdayelise commented:
@elizaveta239 if it's ok with you I could fix them - it's pretty minor and I think it's better to get PRs in rather than have them lingering for ages. But if you really want to tackle them I can leave it for you :)

Elizaveta239 commented:
@pfctdayelise Hi! I'm online again :) I added a small fix for passing your new tests and fixed one old test. But authomatic checks have failed and I can't understand why.

RonnyPfannschmidt commented:
@Elizaveta239 can you add another commit or perhaps even rebase to master, it seems the checks refuse to run due to automated merging isn't working in something like the changelog

Elizaveta239 commented:
@pfctdayelise I merged master to my branch and now authomatic checks are passed. 
Also I wanted to ask you a question about links to commits you mentioned in your answer. They are shown in pytest-dev repository, but I can't find them in a history. Where are they situated?

Elizaveta239 commented:
@RonnyPfannschmidt I merged master to my branch and it helped. Thank you! :)

nicoddemus commented:
> They are shown in pytest-dev repository, but I can't find them in a history. Where are they situated?

I wanted to ask that as well. :smile:

pfctdayelise commented:
@Elizaveta239 I think I checked out your master, but I either didn't fork it first, or the remotes are not set up correctly. So they're kind of floating around in the git ether somewhere!

pfctdayelise commented:
Is it possible to merge 36b86af into this? I don't mind even if it's done as a patch.

RonnyPfannschmidt commented:
@pfctdayelise just do a pull request against @Elizaveta239 's branch?

nicoddemus commented:
> @pfctdayelise just do a pull request against @Elizaveta239 's branch?

Agree, then when @pfctdayelise merges here own branch this PR will be also merged automatically. :blush:
